jax/tree_util.py

```diff
@@ -152,7 +159,7 @@ def tree_unflatten(cls, aux_data, children):
   register_pytree_node(cls, op.methodcaller('tree_flatten'), cls.tree_unflatten)
   return cls
 
-def tree_map(f, tree):
+def tree_map(f: Callable[[Any], Any], tree: T) -> T:
```

Comment:
I don't think it's correct to conclude that the input and output are the same type.
Counterexample:
```
jax.tree_util.tree_map(lambda x: "hi", 3)
```

Same several other places in this file.

Reply:
Oh, great point!  I'll fix this.

Reply:
With the new overloads, we get
```python
reveal_type(tree_map(lambda x: "hi", 3))  # Revealed type is 'builtins.int'
reveal_type(tree_map(lambda x: 2 * x, [3.0, 4.0]))  # Revealed type is 'builtins.list[builtins.float]'
reveal_type(tree_multimap(f, [4.5, 3.2], [3.0, 4.0]))  # Revealed type is 'Union[builtins.list[builtins.float], builtins.float*]'
reveal_type(tree_multimap(g, 4.5, 2))  # Revealed type is 'Union[builtins.float, builtins.str*]'
```

Reply:
@hawkinsp Does it look okay to you now?